Update k8s tests and libraries to v1.27.0

tklauser opened the pull request:
Updating k8s libs to v1.27.0 requires updating sigs.k8s.io/controller-runtime to an unreleased version with https://github.com/kubernetes-sigs/controller-runtime/pull/2223 in place. This in turn requires some additional adjustments to changed APIs in the controller-runtime, mainly in operator/pkg/gateway-api.

The service.kubernetes.io/topology-aware-hints annotation got deprecated with k8s 1.27 in favor of service.kubernetes.io/topology-mode (see https://github.com/kubernetes/kubernetes/pull/116522). This change keeps support for the deprecated annotation. Support for the new annotation will be added in a successive commit.

Supersedes #24553

```release-note
Add support for Kubernetes v1.27
```

aanm reviewed (changes requested):
Discussed offline

zacharysarah commented on vendor/github.com/golang-jwt/jwt/v4/README.md:
```diff
@@ -36,19 +36,41 @@ The part in the middle is the interesting bit.  It's called the Claims and conta
 
+## Installation Guidelines
+
+1. To install the jwt package, you first need to have [Go](https://go.dev/doc/install) installed, then you can use the command below to add `jwt-go` as a dependency in your Go program.
```
Edits for clarity
```suggestion
1. To install the jwt package, first install [Go](https://go.dev/doc/install). Then you can use the following command to add `jwt-go` as a dependency in your Go program.
```

tklauser replied:
@zacharysarah This change is in a 3rd party dependency package which we copy 1:1 to our code base (i.e. everything in `vendor/`). We'd need to make the changes upstream first and then vendor them back in. I think this is out of scope for this PR.

zacharysarah commented on vendor/github.com/golang-jwt/jwt/v4/README.md:
```diff
@@ -36,19 +36,41 @@ The part in the middle is the interesting bit.  It's called the Claims and conta
 ## Extensions
 
-This library publishes all the necessary components for adding your own signing methods.  Simply implement the `SigningMethod` interface and register a factory method using `RegisterSigningMethod`.  
+This library publishes all the necessary components for adding your own signing methods or key functions.  Simply implement the `SigningMethod` interface and register a factory method using `RegisterSigningMethod` or provide a `jwt.Keyfunc`.
```
Avoid "simply", "just", "easy", etc.
```suggestion
This library publishes all the necessary components for adding your own signing methods or key functions.  Implement the `SigningMethod` interface and register a factory method using `RegisterSigningMethod` or provide a `jwt.Keyfunc`.
```

tklauser replied:
Same here, these docs are part of a 3rd party dependency.

zacharysarah commented on vendor/github.com/golang-jwt/jwt/v4/README.md:
```diff
@@ -36,19 +36,41 @@ The part in the middle is the interesting bit.  It's called the Claims and conta
-This library publishes all the necessary components for adding your own signing methods.  Simply implement the `SigningMethod` interface and register a factory method using `RegisterSigningMethod`.  
+This library publishes all the necessary components for adding your own signing methods or key functions.  Simply implement the `SigningMethod` interface and register a factory method using `RegisterSigningMethod` or provide a `jwt.Keyfunc`.
+
+A common use case would be integrating with different 3rd party signature providers, like key management services from various cloud providers or Hardware Security Modules (HSMs) or to implement additional standards.
```
Edits for clarity
```suggestion
A common use case is to integrate with third party signature providers, like key management services from various cloud providers or hardware security modules (HSMs), or to implement additional standards.
```

tklauser replied:
Same here, these docs are part of a 3rd party dependency.

zacharysarah commented on vendor/github.com/golang-jwt/jwt/v4/README.md:
```diff
@@ -36,19 +36,41 @@ The part in the middle is the interesting bit.  It's called the Claims and conta
+| AWS       | Integrates with AWS Key Management Service, KMS                                                          | https://github.com/matelang/jwt-go-aws-kms |
+| JWKS      | Provides support for JWKS ([RFC 7517](https://datatracker.ietf.org/doc/html/rfc7517)) as a `jwt.Keyfunc` | https://github.com/MicahParks/keyfunc       |
+
+*Disclaimer*: Unless otherwise specified, these integrations are maintained by third parties and should not be considered as a primary offer by any of the mentioned cloud providers
```
Use available RST callouts
```suggestion
.. note::
Unless otherwise specified, these integrations are maintained by third parties and should not be considered as a primary offer by any of the mentioned cloud providers
```

tklauser replied:
Same here, these docs are part of a 3rd party dependency.

zacharysarah commented on vendor/github.com/golang-jwt/jwt/v4/SECURITY.md:
```diff
@@ -0,0 +1,19 @@
+
+## Supported Versions
+
+As of February 2022 (and until this document is updated), the latest version `v4` is supported.
```
Edited for plain language
```suggestion
As of February 2022, the latest version supported is `v4`.
```

tklauser replied:
Same here, these docs are part of a 3rd party dependency.

zacharysarah commented on vendor/github.com/golang-jwt/jwt/v4/SECURITY.md:
```diff
@@ -0,0 +1,19 @@
+
+## Reporting a Vulnerability
+
+If you think you found a vulnerability, and even if you are not sure, please report it to jwt-go-security@googlegroups.com or one of the other [golang-jwt maintainers](https://github.com/orgs/golang-jwt/people). Please try be explicit, describe steps to reproduce the security issue with code example(s).
+
+You will receive a response within a timely manner. If the issue is confirmed, we will do our best to release a patch as soon as possible given the complexity of the problem.
+
+## Public Discussions
+
+Please avoid publicly discussing a potential security vulnerability.
+
+Let's take this offline and find a solution first, this limits the potential impact as much as possible.
+
+We appreciate your help!
```
Edit for clarity
```suggestion
If you think you've found a vulnerability, even if you're not sure, please report it to jwt-go-security@googlegroups.com or one of the other [golang-jwt maintainers](https://github.com/orgs/golang-jwt/people). Be explicit when describing steps to reproduce the security issue. Provide code example(s).

You'll receive a response as soon as we can get back to you. If the issue is confirmed, we'll do our best to release a patch as soon as possible.

## Public Discussions

Please avoid discussing potential security vulnerabilities in public.

Finding a solution first limits a vulnerability's potential impact.

We appreciate your help!
```

tklauser replied:
Same here, these docs are part of a 3rd party dependency.

zacharysarah reviewed (changes requested):
@tklauser 👋🏻 Some edits for clarity, otherwise LGTM from a docs perspective

tklauser committed:
Update k8s tests and libraries to v1.27.0

Updating k8s libs to v1.27.0 requires updating
sigs.k8s.io/controller-runtime to an unreleased version with
https://github.com/kubernetes-sigs/controller-runtime/pull/2223 in
place. This in turn requires some additional adjustments to changed APIs
in the controller-runtime, mainly in operator/pkg/gateway-api.

The service.kubernetes.io/topology-aware-hints annotation got deprecated
with k8s 1.27 in favor of service.kubernetes.io/topology-mode (see
https://github.com/kubernetes/kubernetes/pull/116522). This change keeps
support for the deprecated annotation. Support for the new annotation
will be added in a successive commit.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
k8s: add support for service.kubernetes.io/topology-mode

k8s 1.27 deprecated the service.kubernetes.io/topology-aware-hints
annotation in favor of the service.kubernetes.io/topology-mode
annotation to allow implementation-specific values for the annotation.
See https://github.com/kubernetes/kubernetes/pull/116522 for more
details.

Add support for service.kubernetes.io/topology-mode, letting the
deprecated service.kubernetes.io/topology-aware-hints take precedence.
This is in correspondence with the upstream implementation.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

sayboras reviewed (approved):
Thanks and looks good to me for servicemesh related files
